### ISSUE TITLE
Listing with webdav not working with NextCloud external storages

### ISSUE THREAD

**Issue by buergi** [ncw/rclone#2420]
When listing a webdav (NextCloud) remote I get the following error

```
rclone ls myremote:
2018/07/27 23:44:13 ERROR : : error listing: couldn't list files: parsing time "" as "Mon Jan _2 15:04:05 MST 2006": cannot parse "" as "Mon"
2018/07/27 23:44:13 Failed to ls: couldn't list files: parsing time "" as "Mon Jan _2 15:04:05 MST 2006": cannot parse "" as "Mon"
```

I tried the current release v1.42 and also a build from the current revision 9c90b5e both on Windows 10 with amd64 architecture.
I found out the problem is that I have an external storage configured in NextCloud and the corresponding file entry has no modification time:

```
  <d:response>
    <d:href>/remote.php/webdav/ftp/</d:href>
    <d:propstat>
      <d:prop>
        <d:resourcetype>
          <d:collection/>
        </d:resourcetype>
        <d:quota-used-bytes>5</d:quota-used-bytes>
        <d:quota-available-bytes>-3</d:quota-available-bytes>
        <d:getetag>"5581aa796f9be"</d:getetag>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
    <d:propstat>
      <d:prop>
        <d:getlastmodified/>
        <d:getcontentlength/>
        <d:getcontenttype/>
      </d:prop>
      <d:status>HTTP/1.1 404 Not Found</d:status>
    </d:propstat>
  </d:response>
```

This bug is somewhat related to #1952, however the problem here is not the wrong time format but the missing modification time.
I can confirm that it works with the following dirty hack in UnmarshalXML in webdav/api/types.go

```
if (v=="") {
	*t = Time(time.Now())
	return nil
}
```

However, I think the proper way of solving this might be evaluating the d:status tag and see if it is OK before parsing the properties.

**Comment by ncw** [comment]
> However, I think the proper way of solving this might be evaluating the d:status tag and see if it is OK before parsing the properties.

I don't understand why the example you posted has two `d:status` fields, one OK and one not found?

The way the XML parsing works (it parses all the input at once), it might be best to do something very similar to what you suggested so just set the current time if the input time was empty.

Does that then work for you? Or does the status not found confuse it?

If so I suggest you send a pull request with that in.

**Comment by fifh** [comment]
Hello. Same error for me, with a context slightly different : I'm using Nextcloud (13) with an object store (Ceph / S3) as primary storage.

And I confirm that the hack of @buergi works. Thank you !

**Comment by buergi** [comment]
I guess what is meant by the two propsets is that some properties are ok and others could not be retrieved. I don't know if this is in the standard or not.

We can use my suggestion as a hack, however, I'm not sure if this couldn't cause trouble when syncing. If the date is set to the current date when the actual date could not be retrieved this could maybe cause conflicts with real changes, or at least unnecessary syncing. 
What do you think?

**Comment by ncw** [comment]
I think your suggested fix is a good one. There is no time so rclone returning the current time is the correct thing to do.

I'm not sure what the 404 status is doing there, maybe it means that the date isn't available. Either way the object does exist doesn't it?

This could potentially cause excess syncing. This is easy to work around with the --size-only flag though.

However your workaround is a soft fail which means that things will carry on copying which is good user experience.

Perhaps we should warn the user once that the remote isn't returning modtimes. I probably wouldn't bother though as there are many other places in the rclone backends which do exactly the same thing. Maybe a debug would be a good idea.

**Comment by buergi** [comment]
The question is shall files with missing modtimes be skipped or forcefully copied (which is what setting modtime to now actually means, does it?). I would say skipping would be a better failsafe action than copying and maybe overwriting other changes is it?
I myself prefer not synced files over unintentionally overwritten files.

Maybe the best possible way to react is to ignore the modtime for items which don't have a valid one and rely on other attributes and if not sure not better not to sync than to overwrite something.

**Comment by ncw** [comment]
> The question is shall files with missing modtimes be skipped or forcefully copied (which is what setting modtime to now actually means, does it?). I would say skipping would be a better failsafe action than copying and maybe overwriting other changes is it?
I myself prefer not synced files over unintentionally overwritten files.

I see what you mean.

> Maybe the best possible way to react is to ignore the modtime for items which don't have a valid one and rely on other attributes and if not sure not better not to sync than to overwrite something

rclone doesn't have a concept of an individual modtime being invalid unfortunately.

The rclone way of doing this is to change the Precision of the backend from 1 second to `fs.ModTimeNotSupported`.  This tells the rest of rclone to ignore the mod times and use other things (like the size) to do the sync.

Resetting the Precision for the backend would be a good thing to do if **all** the files don't have a modtime.  Is that the case?

**Comment by zarbis** [comment]
@ncw I think I'm facing more general issue of incorrect time returned by server:

> rclone -V
> rclone v1.45
> - os/arch: linux/amd64
> - go version: go1.11.2

> rclone ls :webdav: --webdav-url https://webdav.mail.ru ...
> 2018/11/29 16:43:06 Failed to ls: couldn't list files: parsing time "Mon, 06 Jan 50470 09:13:20 GMT" as "2006-01-02T15:04:05Z07:00": cannot parse "Mon, 06 Jan 50470 09:13:20 GMT" as "2006"

I think it needs same sort of graceful treatment.

**Comment by ncw** [comment]
> @ncw I think I'm facing more general issue of incorrect time returned by server:
> 
> > rclone -V
> > rclone v1.45
> > 
> > * os/arch: linux/amd64
> > * go version: go1.11.2
> 
> > rclone ls :webdav: --webdav-url https://webdav.mail.ru ...
> > 2018/11/29 16:43:06 Failed to ls: couldn't list files: parsing time "Mon, 06 Jan 50470 09:13:20 GMT" as "2006-01-02T15:04:05Z07:00": cannot parse "Mon, 06 Jan 50470 09:13:20 GMT" as "2006"
> 
> I think it needs same sort of graceful treatment.

Maybe the server is time travelling into the year 50470 ;-)

Here is the code.

https://github.com/ncw/rclone/blob/50452207d912447197c149cb3aa96ae7c26ef838/backend/webdav/api/types.go#L143-L174

What would you suggest - logging an error (with `fs.Errorf`) or debug and returning the epoch time?

Fancy sending a pull request?

**Comment by zarbis** [comment]
@ncw my problem is not in parsing absurd date like year 50470, but being unable to do `ls/sync/etc` when dates are broken. Especially when dates are not used like in `ls` or `sync --size-only`.

**Comment by ncw** [comment]
> my problem is not in parsing absurd date like year 50470, but being unable to do `ls/sync/etc` when dates are broken. Especially when dates are not used like in `ls` or `sync --size-only`.

Sure I understand.  However dates are parsed regardless of whether they are used!  So the suggested fix above would fix your problem I think.

**Comment by zarbis** [comment]
@ncw returning the epoch with a warning without hard-failing is what would I consider most sensible in case of non-parseable date.

**Comment by ncw** [comment]
@zarbis can you give this a go?  It should fail softly with a single error message and return the epoch.

https://beta.rclone.org/branch/v1.45-010-g0410c2c8-fix-webdav-times-beta/ (uploaded in 15-30 mins)

**Comment by zarbis** [comment]
@ncw I've retried `sync` on same src and destination with new binary and it seems to finish while printing warning:
```
2018-11-30 14:32:15 ERROR : Failed to parse time "Mon, 06 Jan 50470 09:13:20 GMT" - using the epoch
Transferred:      703.449k / 703.449 kBytes, 100%, 27.905 kBytes/s, ETA 0s
Errors:                 0
Checks:             10195 / 10195, 100%
Transferred:            4 / 4, 100%
Elapsed time:       25.2s
```

**Comment by zarbis** [comment]
@ncw any info on when will be the release including this fix?

**Comment by ncw** [comment]
Oops, sorry I forgot to merge it!

I've merged it now - it will be in [the latest beta](https://beta.rclone.org) in 30 mins or so and be released with v1.46

**Comment by zarbis** [comment]
@ncw any info on 1.46 release date?
